config/application.rb

```diff
@@ -47,5 +47,13 @@ class Application < Rails::Application
     config.action_dispatch.rack_cache = nil
 
     config.paths["log"] = ENV["LOG_PATH"] if ENV["LOG_PATH"]
+
+    # Slimmer is used by the govuk_publishing_components gem, however it inserts itself
+    # automatically as middleware in the host Rails application
+    # Disable Slimmer middleware for Static and enable for use in gem only
+    if defined?(GovukPublishingComponents) && defined?(Slimmer)
```

Comment:
I don't have a great understanding of Railties - is this config always guaranteed to be run after the Slimmer railtie has injected itself into the middleware? If so, why do we need to check `defined?(Slimmer)`? (and `defined?(GovukPublishingComponents)` at that rate).

Reply:
We need to check that Slimmer is defined in order to access `Slimmer::App`
We need to know if GovukPublishingComponents is defined to add the slimmer middleware to the engine.

There isn't an easy way to query if some middleware exists.

Reply:
Yes, but what I mean is - shouldn't we _ensure_ we run this after `Slimmer` has been railtied in? If it's not, then surely the component guide won't work properly? The fact that you have to check if it's been loaded is a smell that says we don't understand the load order properly, which probably isn't great.

Reply:
We require the gems before configuring the app:
https://github.com/alphagov/static/blame/use-guide-gem/config/application.rb#L9

The publishing components gem requires slimmer which adds the railtie to the host app. The railtie must be loaded during the Rails boot process.

https://github.com/alphagov/govuk_publishing_components/blob/0ffeab9ebe1fd53300c80bea63572ed39b6bf18a/lib/govuk_publishing_components/engine.rb#L5
https://github.com/alphagov/slimmer/blob/868095a99fdf559cdad80b9102d882322214bc60/lib/slimmer/railtie.rb

When reaching this logic, if the gem has been required (ie dev/test/on heroku), then the two modules will be defined and we should move the middleware.

The load order looks correct to me?

My interpretation is that we can't run the Slimmer initialisation any later than this.

Reply:
Sorry, I misunderstood again. So Slimmer will be loaded automatically by the GovukPublishingComponent gem, but that's only included in the dev or Heroku environment.

That being so, do we need to check `defined?(Slimmer)`? Isn't that implicit in the check for `defined?(GovukPublishingComponents)`?

Reply:
Strictly no – I left it in as protection against a version change that later removes Slimmer (or its specific `require` call)

Reply:
I personally find it a bit confusing - it makes me wonder when Slimmer wouldn't be defined. If it's always going to be defined, I think we shouldn't check. And - especially given that this guide isn't public-facing - it is probably nicer for it to fail loudly than quietly on the off-chance that Slimmer is removed from the gem.

Reply:
I've updated and removed the Slimmer part of the logic. :+1: